cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -194,47 +190,55 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
             .at(constraint.submap_id.submap_index)
             .data(),
         C_nodes.at(constraint.node_id.trajectory_id)
-            .at(constraint.node_id.node_index -
-                trajectory_data_.at(constraint.node_id.trajectory_id)
-                    .num_trimmed_nodes)
+            .at(constraint.node_id.node_index)
             .data());
   }
 
   // Add penalties for violating odometry or changes between consecutive scans
   // if odometry is not available.
   for (size_t trajectory_id = 0; trajectory_id != node_data_.size();
        ++trajectory_id) {
-    for (size_t node_data_index = 1;
-         node_data_index < node_data_[trajectory_id].size();
-         ++node_data_index) {
-      const bool odometry_available =
-          trajectory_id < odometry_data_.size() &&
-          odometry_data_[trajectory_id].Has(
-              node_data_[trajectory_id][node_data_index].time) &&
-          odometry_data_[trajectory_id].Has(
-              node_data_[trajectory_id][node_data_index - 1].time);
-      const transform::Rigid3d relative_pose =
-          odometry_available
-              ? odometry_data_[trajectory_id]
-                        .Lookup(
-                            node_data_[trajectory_id][node_data_index - 1].time)
-                        .inverse() *
-                    odometry_data_[trajectory_id].Lookup(
-                        node_data_[trajectory_id][node_data_index].time)
-              : transform::Embed3D(
-                    node_data_[trajectory_id][node_data_index - 1]
-                        .initial_point_cloud_pose.inverse() *
-                    node_data_[trajectory_id][node_data_index]
-                        .initial_point_cloud_pose);
-      problem.AddResidualBlock(
-          new ceres::AutoDiffCostFunction<SpaCostFunction, 3, 3, 3>(
-              new SpaCostFunction(Constraint::Pose{
-                  relative_pose,
-                  options_.consecutive_scan_translation_penalty_factor(),
-                  options_.consecutive_scan_rotation_penalty_factor()})),
-          nullptr /* loss function */,
-          C_nodes[trajectory_id][node_data_index - 1].data(),
-          C_nodes[trajectory_id][node_data_index].data());
+    if (node_data_[trajectory_id].size() > 1) {
```

Comment:
Isn't this check redundant? Remove?

Reply:
No, it isn't. `++node_it` causes a problem when `node_data_[trajectory_id]` is empty. For loop doesn't have an initial condition check before `++node_it`

Reply:
I will replace it with `if(!node_data_[trajectory_id].empty())`